File: library/src/main/java/com/avigation/view/NavigationView.java
package com.avigation.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.viewpager.widget.ViewPager;

import com.avigation.adapter.FragmentViewPagerAdapter;
import com.avigation.listener.OnTabItemSelectedListener;

import java.util.ArrayList;
import java.util.List;

/**
 * @WYU-WIN
 * @date 2020/7/14 0014.
 * description：
 */
public class NavigationView extends View implements ItemController {
    //////////////////////////////////////////////////
    //数据准备
    //////////////////////////////////////////////////
    private int containerId;
    private List<Class> fragmentClassList = new ArrayList<>();
    private List<String> titleList = new ArrayList<>();
    private List<Integer> defaultIconResList = new ArrayList<>();
    private List<Integer> selectIconResList = new ArrayList<>();
    private List<Fragment> fragmentList = new ArrayList<>();
    private List<Boolean> roundList = new ArrayList<>();
    private List<Bitmap> defaultBitmapList = new ArrayList<>();
    private List<Bitmap> selectBitmapList = new ArrayList<>();
    private List<Rect> iconRectList = new ArrayList<>();
    private List<Integer> textXList = new ArrayList<>();

    private int defaultColor = Color.parseColor("#999999");
    private int selectColor = Color.parseColor("#ff5d5e");
    private int backgroudColor = Color.parseColor("#FFFFFF");

    private OnTabItemSelectedListener selectedListener;
    private NavigationPager navigationPager;

    private Paint paint = new Paint();

    private int itemCount;
    private int currentSelectIndex = 0;
    private int defaultSelectIndex = 0;
    private int textSize = 12;
    private int iconWidth = 30;
    private int iconHeight = 30;
    private int titleIconMargin = 5;
    private Context context;

    public NavigationView(Context context) {
        super(context, null);
        this.context = context;
    }

    public NavigationView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
    }

    @Override
    public NavigationView setContainer(int containerId) {
        this.containerId = containerId;
        return this;
    }

    @Override
    public NavigationView setBackgroudColor(int backgroudColor) {
        this.backgroudColor = backgroudColor;
        return this;
    }

    @Override
    public NavigationView setColors(int defaultColor, int selectColor) {
        this.defaultColor = defaultColor;
        this.selectColor = selectColor;
        return this;
    }

    @Override
    public NavigationView setTitleSize(int textSize) {
        this.textSize = textSize;
        return this;
    }

    @Override
    public NavigationView setIconWidth(int iconWidth) {
        this.iconWidth = iconWidth;
        return this;
    }

    @Override
    public NavigationView setTitleIconMargin(int titleIconMargin) {
        this.titleIconMargin = titleIconMargin;
        return this;
    }

    @Override
    public NavigationView setIconHeight(int iconHeight) {
        this.iconHeight = iconHeight;
        return this;
    }

    @Override
    public NavigationView addRoundItem(Class fragmentClass, String title, int iconResBefore, int iconResAfter) {
        fragmentClassList.add(fragmentClass);
        titleList.add(title);
        defaultIconResList.add(iconResBefore);
        selectIconResList.add(iconResAfter);
        roundList.add(true);
        return this;
    }

    @Override
    public NavigationView addItem(Class fragmentClass, String title, int iconResBefore, int iconResAfter) {
        fragmentClassList.add(fragmentClass);
        titleList.add(title);
        defaultIconResList.add(iconResBefore);
        selectIconResList.add(iconResAfter);
        roundList.add(false);
        return this;
    }

    @Override
    public NavigationView setDefaultSelect(int selectIndex) {//从0开始
        this.defaultSelectIndex = selectIndex;
        return this;
    }

    @Override
    public void setSelect(int index) {
        if (navigationPager == null || index >= fragmentList.size() || index < 0) {
            return;
        }

        if (navigationPager.enableSlide()) {
            navigationPager.setCurrentItem(target);
        } else {
            switchFragment(target);
        }
        target = index;
        currentSelectIndex = index;
        invalidate();
    }

    @Override
    public Fragment getFragmentByIndex(int index) {
        if (index >= fragmentList.size()) {
            return null;
        }
        return fragmentList.get(index);
    }

    @Override
    public int getItemCount() {
        return fragmentList.size();
    }

    @Override
    public String getItemTitle(int index) {
        if (index >= titleList.size()) {
            return null;
        }
        return titleList.get(index);
    }

    @Override
    public NavigationPager build() {
        itemCount = fragmentClassList.size();
        //预创建icon的Rect并缓存
        for (int i = 0; i < itemCount; i++) {
            Bitmap beforeBitmap = getBitmap(defaultIconResList.get(i));
            defaultBitmapList.add(beforeBitmap);

            Bitmap afterBitmap = getBitmap(selectIconResList.get(i));
            selectBitmapList.add(afterBitmap);

            Rect rect = new Rect();
            iconRectList.add(rect);

            Class clx = fragmentClassList.get(i);
            try {
                Fragment fragment = (Fragment) clx.newInstance();
                fragmentList.add(fragment);
            } catch (InstantiationException | IllegalAccessException e) {
                e.printStackTrace();
            }
        }

        currentSelectIndex = defaultSelectIndex;
        if (navigationPager == null) {
            navigationPager = new NavigationPager();
        }
        return navigationPager;
    }

    public class NavigationPager {
        private ViewPager mViewPager;
        private ViewPagerPageChangeListener mPageChangeListener;
        private FragmentViewPagerAdapter mViewPagerAdapter;
        private boolean enableSlide = false;

        public NavigationPager setWithViewPager(ViewPager viewPager) {
            if (viewPager == null) {
                enableSlide = false;
                return this;
            }
            mViewPager = viewPager;
            mViewPagerAdapter = new FragmentViewPagerAdapter(((AppCompatActivity) context).getSupportFragmentManager(), fragmentList);
            mViewPager.setAdapter(mViewPagerAdapter);

            if (mPageChangeListener != null) {
                mViewPager.removeOnPageChangeListener(mPageChangeListener);
            } else {
                mPageChangeListener = new ViewPagerPageChangeListener();
            }

            mViewPager.addOnPageChangeListener(mPageChangeListener);
            enableSlide = true;
            return this;
        }

        public boolean enableSlide() {
            return enableSlide;
        }

        public void setCurrentItem(int item) {
            if (mViewPager != null) {
                mViewPager.setCurrentItem(item);
            }
        }

        public void applay() {
            if (mViewPager == null) {
                switchFragment(currentSelectIndex);
            }
            invalidate();
        }
    }


    @Override
    public void setTabItemSelectedListener(OnTabItemSelectedListener listener) {
        selectedListener = listener;
    }

    private class ViewPagerPageChangeListener implements ViewPager.OnPageChangeListener {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

        }

        @Override
        public void onPageSelected(int position) {
            if (selectedListener != null) {
                selectedListener.onSelected(position, currentSelectIndex);
            }
            setSelect(position);
        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }
    }

    //////////////////////////////////////////////////
    //初始化数据基础
    //////////////////////////////////////////////////
    private int titleBaseLine;
    private int marginTop = 0;
    private int parentItemWidth;

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (itemCount != 0) {
            //单个item宽高
            parentItemWidth = getWidth() / itemCount;
            int parentItemHeight = getHeight();
            if (hadRoudItem()) {
                marginTop = 16;
            } else {
                marginTop = 0;
            }

            //图标边长
            int iconWidth = dp2px(this.iconWidth);//先指定20dp
            int iconHeight = dp2px(this.iconHeight);
            //图标文字margin
            int textIconMargin = dp2px(titleIconMargin);
            //标题高度
            int titleSize = dp2px(textSize);//这里先指定10dp
            paint.setTextSize(titleSize);
            Rect measureRect = new Rect();
            String textBound = "测量高度";
            paint.getTextBounds(textBound, 0, textBound.length(), measureRect);

            //从而计算得出图标的起始top坐标、文本的baseLine
            titleBaseLine = getHeight() - textIconMargin - measureRect.height() - measureRect.top;
            int iconTop = marginTop + textIconMargin;

            //对icon的rect的参数进行赋值
            int firstRectX = (parentItemWidth - iconWidth) / 2;//第一个icon的左
            for (int i = 0; i < itemCount; i++) {
                int rectX = i * parentItemWidth + firstRectX;
                boolean isRoundItem = roundList.get(i);
                if (isRoundItem) {
                    String titleText = titleList.get(i);
                    if (TextUtils.isEmpty(titleText)) {
                        // 计算有弧度，但是没有文字的Item的Rect
                        int rleft = rectX - marginTop / 2;
                        int rtop = marginTop - marginTop / 2;
                        int rectWh = getHeight() - (marginTop - marginTop / 2) * 2;
                        int rright = rleft + rectWh;
                        int rbottom = rtop + rectWh;

                        Rect temp = iconRectList.get(i);
                        temp.left = rleft;
                        temp.top = rtop;
                        temp.right = rright;
                        temp.bottom = rbottom;
                    } else {
                        // 计算有弧度，并且有文字的Item的Rect
                        Rect temp = iconRectList.get(i);
                        temp.left = rectX - marginTop / 2;
                        temp.top = marginTop - marginTop / 2;
                        temp.right = rectX + iconWidth + marginTop / 2;
                        temp.bottom = marginTop + iconHeight + marginTop / 2;
                    }
                } else {
                    // 计算没有弧度的Item的Rect
                    Rect temp = iconRectList.get(i);
                    temp.left = rectX;
                    temp.top = iconTop;
                    temp.right = rectX + iconWidth;
                    temp.bottom = iconTop + iconHeight;
                }

                //计算Item文字位置
                String titleText = titleList.get(i);
                // 测量文字宽度
                paint.getTextBounds(titleText, 0, titleText.length(), measureRect);
                textXList.add((parentItemWidth - measureRect.width()) / 2 + parentItemWidth * i);
            }


        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (itemCount != 0) {
            paint.setAntiAlias(true);
            paint.setColor(backgroudColor);
            // 画背景颜色
            if (hadRoudItem()) {
                canvas.drawRect(0, marginTop, getWidth(), getHeight(), paint);
            } else {
                canvas.drawRect(0, 0, getWidth(), getHeight(), paint);
            }

            for (int i = 0; i < itemCount; i++) {
                paint.setColor(backgroudColor);
                boolean isRoundItem = roundList.get(i);
                if (isRoundItem) {
                    // 画弧形区域
                    Rect rect = iconRectList.get(i);
                    int radius = getHeight() / 2;
                    canvas.drawCircle(rect.centerX(), radius, radius, paint);
                }
                //画对应图标
                Bitmap bitmap = null;
                if (i == currentSelectIndex) {
                    bitmap = selectBitmapList.get(i);
                } else {
                    bitmap = defaultBitmapList.get(i);
                }
                Rect rect = iconRectList.get(i);
                canvas.drawBitmap(bitmap, null, rect, paint);//null代表bitmap全部画出
                //画文字
                String title = titleList.get(i);
                if (i == currentSelectIndex) {
                    paint.setColor(selectColor);
                } else {
                    paint.setColor(defaultColor);
                }
                int x = textXList.get(i);
                canvas.drawText(title, x, titleBaseLine, paint);
            }
        }
    }

    //////////////////////////////////////////////////
    //点击事件: 通过点的坐标所在区域判断应该响应哪一个Fragment
    //////////////////////////////////////////////////

    int target = -1;
    long clickTime = 0;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                target = controlArea((int) event.getX());
                // 可以在这里改造长按事件
                break;
            case MotionEvent.ACTION_UP:
                if (event.getY() < 0) {
                    break;
                }
                if (target == controlArea((int) event.getX())) {
                    //这里触发点击事件
                    long currentTime = System.currentTimeMillis();
                    if (currentSelectIndex == target && selectedListener != null && (currentTime - clickTime) < 500) {
                        selectedListener.onRepeat(currentSelectIndex);
                    }
                    clickTime = System.currentTimeMillis();
                    setSelect(target);
                }
                target = -1;
                break;
        }
        return true;
    }

    ///////////////////////////////////////////////////////
    // 内部处理方法区域
    ///////////////////////////////////////////////////////

    private Bitmap getBitmap(int resId) {
        BitmapDrawable bitmapDrawable = (BitmapDrawable) context.getResources().getDrawable(resId);
        return bitmapDrawable.getBitmap();
    }

    private int dp2px(float dpValue) {
        float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * 是否是弧形Item
     *
     * @return
     */
    private boolean hadRoudItem() {
        for (int i = 0; i < itemCount; i++) {
            if (roundList.get(i)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 返回点位置所在区域
     *
     * @param x
     * @return
     */
    private int controlArea(int x) {
        return x / parentItemWidth;
    }

    //////////////////////////////////////////////////
    //碎片处理代码
    //////////////////////////////////////////////////
    private Fragment currentFragment;

    private void switchFragment(int target) {
        if (target == currentSelectIndex || target < 0 || target >= fragmentList.size()) {
            return;
        }
        Fragment fragment = fragmentList.get(target);
        int frameLayoutId = containerId;
        if (fragment != null) {
            //只支持AppCompatActivity
            FragmentTransaction transaction = ((AppCompatActivity) context).getSupportFragmentManager().beginTransaction();
            if (fragment.isAdded()) {
                if (currentFragment != null) {
                    transaction.hide(currentFragment).show(fragment);
                } else {
                    transaction.show(fragment);
                }
            } else {
                if (currentFragment != null) {
                    transaction.hide(currentFragment).add(frameLayoutId, fragment);
                } else {
                    transaction.add(frameLayoutId, fragment);
                }
            }
            currentFragment = fragment;
            transaction.commit();
        }
    }
}

File: library/src/main/java/com/wang/navigation/NavigationView.kt
package com.wang.navigation

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.graphics.drawable.AdaptiveIconDrawable
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.text.TextUtils
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.wang.navigation.adapter.FragmentViewPagerAdapter
import com.wang.navigation.control.ItemController
import com.wang.navigation.listener.OnTabItemSelectedListener
import java.util.*


/**
 * @WYU-WIN
 * @date 2020/7/14 0014.
 * description：
 */
class NavigationView : View, ItemController {
    private var TAG: String = this::class.java.simpleName

    //////////////////////////////////////////////////
    //数据准备
    //////////////////////////////////////////////////
    private var containerId = 0
    private val fragmentClassList: MutableList<Class<*>> = ArrayList()
    private val titleList: MutableList<String?> = ArrayList()
    private val defaultIconResList: MutableList<Int> = ArrayList()
    private val selectIconResList: MutableList<Int> = ArrayList()
    private val fragmentList: MutableList<Fragment> = ArrayList()
    private val roundList: MutableList<Boolean> = ArrayList()
    private val iconRectList: MutableList<Rect> = ArrayList()
    private val textXList: MutableList<Int> = ArrayList()
    private var defaultColor = Color.parseColor("#999999")
    private var selectColor = Color.parseColor("#ff5d5e")
    private var backgroudColor = Color.parseColor("#FFFFFF")
    private var selectedListener: OnTabItemSelectedListener? = null
    private var navigationPager: NavigationPager? = null
    private val paint = Paint()
    private var itemCount = 0
    private var currentSelectIndex = -1
    private var defaultSelectIndex = 0
    private var textSize = 12
    private var iconWidth = 30
    private var iconHeight = 30
    private var titleIconMargin = 5

    constructor(context: Context) : super(context, null) {
    }

    constructor(
        context: Context,
        attrs: AttributeSet?
    ) : super(context, attrs) {
    }

    override fun setContainer(containerId: Int): NavigationView {
        this.containerId = containerId
        return this
    }

    override fun setBackgroudColor(backgroudColor: Int): NavigationView {
        this.backgroudColor = backgroudColor
        return this
    }

    override fun setColors(defaultColor: Int, selectColor: Int): NavigationView {
        this.defaultColor = defaultColor
        this.selectColor = selectColor
        return this
    }

    override fun setTextSize(textSize: Int): NavigationView {
        this.textSize = textSize
        return this
    }

    override fun setIconWidth(iconWidth: Int): NavigationView {
        this.iconWidth = iconWidth
        return this
    }

    override fun setMargin(titleIconMargin: Int): NavigationView {
        this.titleIconMargin = titleIconMargin
        return this
    }

    override fun setIconHeight(iconHeight: Int): NavigationView {
        this.iconHeight = iconHeight
        return this
    }

    override fun addRoundItem(
        fragmentClass: Class<*>,
        title: String,
        iconResBefore: Int,
        iconResAfter: Int
    ): NavigationView {
        fragmentClassList.add(fragmentClass)
        titleList.add(title)
        defaultIconResList.add(iconResBefore)
        selectIconResList.add(iconResAfter)
        roundList.add(true)
        return this
    }

    override fun addItem(
        fragmentClass: Class<*>,
        title: String,
        iconResBefore: Int,
        iconResAfter: Int
    ): NavigationView {
        fragmentClassList.add(fragmentClass)
        titleList.add(title)
        defaultIconResList.add(iconResBefore)
        selectIconResList.add(iconResAfter)
        roundList.add(false)
        return this
    }

    override fun setDefaultSelect(selectIndex: Int): NavigationView { //从0开始
        defaultSelectIndex = selectIndex
        return this
    }

    override fun setSelect(index: Int) {
        if (index >= fragmentList.size || index < 0) {
            return
        }
        if (navigationPager != null && navigationPager!!.enableSlide()) {
            navigationPager?.setCurrentItem(index)
        } else {
            switchFragment(index)
        }
        invalidate()
    }

    override fun getFragmentByIndex(index: Int): Fragment? {
        return if (index >= fragmentList.size) {
            null
        } else fragmentList[index]
    }

    override fun getItemCount(): Int {
        return fragmentList.size
    }

    override fun getItemTitle(index: Int): String? {
        return if (index >= titleList.size) {
            null
        } else titleList[index]
    }

    fun builder() {
        itemCount = fragmentClassList.size
        //预创建bitmap的Rect并缓存
        //预创建icon的Rect并缓存
        for (i in 0 until itemCount) {
            val rect = Rect()
            iconRectList.add(rect)
            val clx = fragmentClassList[i]
            try {
                val fragment = clx.newInstance() as Fragment
                fragmentList.add(fragment)
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }
        currentSelectIndex = defaultSelectIndex;
        switchFragment(defaultSelectIndex)
        invalidate()
    }

    override fun build(): NavigationPager {
        itemCount = fragmentClassList.size
        //预创建icon的Rect并缓存
        for (i in 0 until itemCount) {

            val rect = Rect()
            iconRectList.add(rect)
            val clx = fragmentClassList[i]
            try {
                val fragment = clx.newInstance() as Fragment
                fragmentList.add(fragment)
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }

        if (navigationPager == null) {
            navigationPager = NavigationPager()
        }
        return navigationPager!!
    }

    inner class NavigationPager {
        private var mViewPager: ViewPager? = null
        private var mPageChangeListener: ViewPagerPageChangeListener? = null
        private var mViewPagerAdapter: FragmentViewPagerAdapter? = null
        private var enableSlide = false
        fun setWithViewPager(viewPager: ViewPager?): NavigationPager {
            if (viewPager == null) {
                enableSlide = false
                return this
            }
            mViewPager = viewPager
            mViewPagerAdapter = FragmentViewPagerAdapter(
                (context as AppCompatActivity).supportFragmentManager,
                fragmentList
            )
            mViewPager!!.adapter = mViewPagerAdapter
            if (mPageChangeListener != null) {
                mViewPager!!.removeOnPageChangeListener(mPageChangeListener!!)
            } else {
                mPageChangeListener = ViewPagerPageChangeListener()
            }
            mViewPager!!.addOnPageChangeListener(mPageChangeListener!!)
            enableSlide = true
            return this
        }

        fun enableSlide(): Boolean {
            return enableSlide
        }

        fun setCurrentItem(item: Int) {
            if (mViewPager != null) {
                mViewPager!!.currentItem = item
            }
        }

        fun apply() {
            currentSelectIndex = defaultSelectIndex;
            if (mViewPager == null) {
                switchFragment(defaultSelectIndex)
            }
            invalidate()
        }
    }

    override fun setTabItemSelectedListener(listener: OnTabItemSelectedListener?) {
        selectedListener = listener
    }

    private inner class ViewPagerPageChangeListener : OnPageChangeListener {
        override fun onPageScrolled(
            position: Int,
            positionOffset: Float,
            positionOffsetPixels: Int
        ) {
        }

        override fun onPageSelected(position: Int) {
            if (selectedListener != null) {
                selectedListener!!.onSelected(position, currentSelectIndex)
            }
            upSelect(position)
        }

        override fun onPageScrollStateChanged(state: Int) {}
    }

    //////////////////////////////////////////////////
    //初始化数据基础
    //////////////////////////////////////////////////
    private var titleBaseLine = 0
    private var marginTop = 0
    private var parentItemWidth = 0
    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        if (itemCount != 0) {
            //单个item宽高
            parentItemWidth = width / itemCount
            val parentItemHeight = height
            marginTop = if (hadRoudItem()) {
                16
            } else {
                0
            }

            //图标边长
            val iconWidth = dp2px(iconWidth.toFloat()) //先指定20dp
            val iconHeight = dp2px(iconHeight.toFloat())
            //图标文字margin
            val textIconMargin = dp2px(titleIconMargin.toFloat())
            //标题高度
            val titleSize = dp2px(textSize.toFloat()) //这里先指定10dp
            paint.textSize = titleSize.toFloat()
            val measureRect = Rect()
            val textBound = "测量高度"
            paint.getTextBounds(textBound, 0, textBound.length, measureRect)

            //从而计算得出图标的起始top坐标、文本的baseLine
            titleBaseLine = height - textIconMargin - measureRect.height() - measureRect.top
            val iconTop = marginTop + textIconMargin

            //对icon的rect的参数进行赋值
            val firstRectX = (parentItemWidth - iconWidth) / 2 //第一个icon的左
            for (i in 0 until itemCount) {
                val rectX = i * parentItemWidth + firstRectX
                val isRoundItem = roundList[i]
                if (isRoundItem) {
                    val titleText = titleList[i]
                    if (TextUtils.isEmpty(titleText)) {
                        // 计算有弧度，但是没有文字的Item的Rect
                        val rleft = rectX - marginTop / 2
                        val rtop = marginTop - marginTop / 2
                        val rectWh = height - (marginTop - marginTop / 2) * 2
                        val rright = rleft + rectWh
                        val rbottom = rtop + rectWh
                        val temp = iconRectList[i]
                        temp.left = rleft
                        temp.top = rtop
                        temp.right = rright
                        temp.bottom = rbottom
                    } else {
                        // 计算有弧度，并且有文字的Item的Rect
                        val temp = iconRectList[i]
                        temp.left = rectX - marginTop / 2
                        temp.top = marginTop - marginTop / 2
                        temp.right = rectX + iconWidth + marginTop / 2
                        temp.bottom = marginTop + iconHeight + marginTop / 2
                    }
                } else {
                    // 计算没有弧度的Item的Rect
                    val temp = iconRectList[i]
                    temp.left = rectX
                    temp.top = iconTop
                    temp.right = rectX + iconWidth
                    temp.bottom = iconTop + iconHeight
                }

                //计算Item文字位置
                val titleText = titleList[i]
                // 测量文字宽度
                paint.getTextBounds(titleText, 0, titleText!!.length, measureRect)
                textXList.add((parentItemWidth - measureRect.width()) / 2 + parentItemWidth * i)
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (itemCount != 0) {
            paint.isAntiAlias = true
            paint.color = backgroudColor
            // 画背景颜色
            if (hadRoudItem()) {
                canvas.drawRect(
                    0f,
                    marginTop.toFloat(),
                    width.toFloat(),
                    height.toFloat(),
                    paint
                )
            } else {
                canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
            }
            for (i in 0 until itemCount) {
                paint.color = backgroudColor
                val isRoundItem = roundList[i]
                if (isRoundItem) {
                    // 画弧形区域
                    val rect = iconRectList[i]
                    val radius = height / 2
                    canvas.drawCircle(
                        rect.centerX().toFloat(),
                        radius.toFloat(),
                        radius.toFloat(),
                        paint
                    )
                }
                //画对应图标
                var bitmap: Bitmap? = null
                val rect = iconRectList[i]
                bitmap = if (i == currentSelectIndex) {
                    getIconBitmap(selectIconResList[i])
                } else {
                    getIconBitmap(defaultIconResList[i])
                }
                if (bitmap != null) {
                    canvas.drawBitmap(bitmap, null, rect, paint) //null代表bitmap全部画出
                }

                //画文字
                val title = titleList[i]
                if (i == currentSelectIndex) {
                    paint.color = selectColor
                } else {
                    paint.color = defaultColor
                }
                val x = textXList[i]
                canvas.drawText(title!!, x.toFloat(), titleBaseLine.toFloat(), paint)
            }
        }
    }

    //////////////////////////////////////////////////
    //点击事件: 通过点的坐标所在区域判断应该响应哪一个Fragment
    //////////////////////////////////////////////////
    var target = -1
    var clickTime: Long = 0

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> target = controlArea(event.x.toInt())
            MotionEvent.ACTION_UP -> {
                if (event.y < 0) {
                    return true
                }
                if (target == controlArea(event.x.toInt())) {
                    //这里触发点击事件
                    val currentTime = System.currentTimeMillis()
                    if (currentSelectIndex == target && selectedListener != null && currentTime - clickTime < 500) {
                        selectedListener!!.onRepeat(currentSelectIndex)
                    }
                    clickTime = System.currentTimeMillis()
                    setSelect(target)
                }
                target = -1
            }
        }
        return true
    }

    ///////////////////////////////////////////////////////
    // 内部处理方法区域
    ///////////////////////////////////////////////////////
    private fun getBitmapFromDrawable(drawable: Drawable): Bitmap {
        val bmp = Bitmap.createBitmap(
            drawable.intrinsicWidth,
            drawable.intrinsicHeight,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bmp)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bmp
    }

    fun getIconBitmap(iconId: Int): Bitmap? {
        return try {
            val icon = ContextCompat.getDrawable(context, iconId)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && icon is AdaptiveIconDrawable) {
                val bitmap = Bitmap.createBitmap(
                    icon.getIntrinsicWidth(),
                    icon.getIntrinsicHeight(),
                    Bitmap.Config.ARGB_8888
                )
                val canvas = Canvas(bitmap)
                icon.setBounds(0, 0, canvas.width, canvas.height)
                icon.draw(canvas)
                bitmap
            } else {
                (icon as BitmapDrawable).bitmap
            }
        } catch (e: Exception) {
            null
        }
    }


    private fun dp2px(dpValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    /**
     * 是否是弧形Item
     *
     * @return
     */
    private fun hadRoudItem(): Boolean {
        for (i in 0 until itemCount) {
            if (roundList[i]) {
                return true
            }
        }
        return false
    }

    /**
     * 返回点位置所在区域
     *
     * @param x
     * @return
     */
    private fun controlArea(x: Int): Int {
        return x / parentItemWidth
    }

    private fun upSelect(index: Int) {
        if (index >= fragmentList.size || index < 0) {
            return
        }
        currentSelectIndex = index
        invalidate()
    }


    //////////////////////////////////////////////////
    //碎片处理代码
    //////////////////////////////////////////////////
    private var currentFragment: Fragment? = null
    private fun switchFragment(pageIndex: Int) {
        if (pageIndex < 0 || pageIndex >= fragmentList.size) {
            return
        }
        currentSelectIndex = target
        val fragment = fragmentList[pageIndex]
        val frameLayoutId = containerId
        if (fragment != null) {
            val transaction =
                (context as AppCompatActivity).supportFragmentManager.beginTransaction()
            if (fragment.isAdded) {
                if (currentFragment != null) {
                    transaction.hide(currentFragment!!).show(fragment)
                } else {
                    transaction.show(fragment)
                }
            } else {
                if (currentFragment != null) {
                    transaction.hide(currentFragment!!).add(frameLayoutId, fragment)
                } else {
                    transaction.add(frameLayoutId, fragment)
                }
            }
            currentFragment = fragment
            transaction.commit()
        }
    }
}